acme client package for pfSense, initial commit

PiBa-NL opened the pull request:
acme client package for pfSense, initial commit

mcfedr commented:
Not sure if its of use, or too late, there is an php implementation of acme thats quite well done, https://github.com/kelunik/acme

wernerdev commented:
Can't you just use the FreeBSD Shell script package as base? http://www.freshports.org/security/letsencrypt.sh/

PiBa-NL commented:
Yes that 'could' be possible now (it didnt exist when i first created the package..), and i'm willing to change it if one of the ESF guy's tells me that would help get the pfSense package committed. Sofar i have not heard anything from them.

vpetersson commented:
+1

razims commented:
+

rbgarga commented:
@PiBa-NL you can go ahead and make all necessary changes. Let me know when you have a final version for proper revision. Thanks!

gonzopancho commented:
@PiBa-NL In case it's not clear, when the changes are made and accepted, we'll be bringing this to pfSense.

idarlund commented:
Take a look at https://github.com/lukas2511/dehydrated - this is the renamed (new) version of the FreeBSD Shell script that @wernerdev referes to.

PiBa-NL commented:
Yes i shortly discussed with rbgarga, i am taking a look at the alternatives to the currently used 'lescript' which seems to hardly be maintained at all.. To which might be the 'best' way to integrate one of the other implementations. t.b.h. ive barely looked at further improvement the past half year as it was uncertain to if it would be 'allowed', the 'go ahead' has been received :) , need a few days though to switch

- kelunik/acme , seems like adding the amp framework just for this is to much complexity/overhead.. evn though the 'php only' implementation is nice..
- lukas2511/dehydrated , its a official freebsd port (the old version anyhow) which i suppose is a pro..
- Neilpang/acme.sh , looks to be most actively developed and integrated with dns update options which could prove nice to have

currently i'm leaning towards the Neilpang implementation.. unless there are compelling reasons ive overlooked the others are way better.?

idarlund commented:
I actually use the DNS update option in acme.sh and it's pretty nice! The good thing about that is that you can still have your haproxy run on port 80 (and 443) without have to turn it off for a cert update. I'd vote for acme.sh if that's the only one who support the DNS update method.

PiBa-NL commented:
@rbgarga, i think its ready to commit, still things to improve i guess but those can follow later once people start testing it

idarlund commented:
DNS-Manual works as expexted! Thanks!
acme/acme_certificates.php is accepting name field with spaces, but the acme script does not wrap the text with quotes. remove space acceptance in php or wrap text.
Last renewed is not updated when running manual "renew" in acme/acme_certificates.php

hoerup commented:
Installed on pfsense 2.3. Works fine although it couldn't find curl out of the box. (tested with DNS-Manual)

heper commented:
@PiBa-NL 
i've tried to make the package in a VM(no clue what i'm doing)
but when doing pkg add on pfSense:
`[2.3.2-RELEASE][root@pfsense.lan]/root: pkg add pfSense-pkg-acme-0.1.txz
Installing pfSense-pkg-acme-0.1...
pkg: Missing dependency 'php56-ftp'
`

i probably did something wrong when creating the package

jim-p commented on security/pfSense-pkg-acme/files/usr/local/pkg/acme/acme_sh.inc:
```diff
@@ -0,0 +1,188 @@
+	function generateAccountKey() {
+		unlink_if_exists("{$this->path_account}/account.key");
+		$this->debug = false;
+		$this->execacmesh("--home {$this->acmeconf} --createAccountKey --accountkeylength 4096 --accountconf {$this->accountconfig}");
```
Must use escapeshellarg() around all parameters involving variables.

jim-p commented on security/pfSense-pkg-acme/files/usr/local/pkg/acme/acme_sh.inc:
```diff
@@ -0,0 +1,188 @@
+	
+	function registeraccount($key) {
+		file_put_contents("{$this->path_account}/account.key", $key);
+		$result = $this->execacmesh(""
```
Must use escapeshellarg() around all parameters involving variables.

PiBa-NL replied:
using escapeshellarg now

jim-p commented on security/pfSense-pkg-acme/files/usr/local/pkg/acme/acme_sh.inc:
```diff
@@ -0,0 +1,188 @@
+		safe_mkdir($certpath);
+		 
+		unlink_if_exists("{$certpath}/{$domain}.key");
+		$this->execacmesh(""
```
Must use escapeshellarg() around all parameters involving variables.

jim-p commented on security/pfSense-pkg-acme/files/usr/local/pkg/acme/acme_sh.inc:
```diff
@@ -0,0 +1,188 @@
+		foreach($domainstosign as $domain) {
+			$domainstr .= " -d {$domain}";
+		}
+		$this->execacmesh(""
```
Must use escapeshellarg() around all parameters involving variables.

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_generalsettings.php:
```diff
@@ -0,0 +1,101 @@
+
+$form->add($section);
+
+print $form;
```
With only this one option and just the checkbox text this page seems really bare. Maybe at least add a little description text about what the renewal job does?
For example, if it uses cron, how often it runs, etc, etc.

PiBa-NL replied:
added a check basically preventing to show the page

jim-p commented on security/pfSense-pkg-acme/files/usr/local/pkg/acme/acme_gui.inc:
```diff
@@ -0,0 +1,139 @@
+		"icon" => "icon_down.gif",
+		"iconsize" => 17),
+	'add' => array(
+		"faicon" => "fa-level-down",
```
Single add buttons should use fa-plus, not fa-level-down. Only use fa-level-down when the page supports adding to the top or bottom of a list (For example, when there is also an fa-level-up add button).

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_accountkeys.php:
```diff
@@ -0,0 +1,281 @@
+	</div>
+	<nav class="action-buttons">
+		<a href="acme_accountkeys_edit.php" role="button" class="btn btn-sm btn-success" title="<?=gettext('Add backend to the end of the list')?>">
+			<i class="fa fa-level-down icon-embed-btn"></i>
```
Single add buttons should use fa-plus, not fa-level-down. Only use fa-level-down when the page supports adding to the top or bottom of a list (For example, when there is also an fa-level-up add button).

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_certificates.php:
```diff
@@ -0,0 +1,378 @@
+	</div>
+	<nav class="action-buttons">
+		<a href="acme_certificates_edit.php" role="button" class="btn btn-sm btn-success" title="<?=gettext('Add backend to the end of the list')?>">
+			<i class="fa fa-level-down icon-embed-btn"></i>
```
Single add buttons should use fa-plus, not fa-level-down. Only use fa-level-down when the page supports adding to the top or bottom of a list (For example, when there is also an fa-level-up add button).

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_certificates_edit.php:
```diff
@@ -0,0 +1,432 @@
+
+$section = new \Form_Section('Edit Certificate options');
+$section->addInput(new \Form_Input('name', 'Name', 'text', $pconfig['name']
+))->setHelp('The name set here will also be used to create or overwrite a certificate that might already exist with this name in the pfSense certificate m.');
```
This sentence is cut off (missing part of a word, or whole words)

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_certificates_edit.php:
```diff
@@ -0,0 +1,432 @@
+	'Acme Account',
+	$pconfig['acmeaccount'],
+	form_name_array($a_accountkeys)
+));
```
It would be helpful if this page would warn the user when there are no account keys, and direct them to create an account key. Someone filling out this form could get this far in before realizing it's impossible to complete the page.

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_certificates_edit.php:
```diff
@@ -0,0 +1,432 @@
+	. "/usr/local/www/.well-known/acme-challenge/<br/>"
+	. "/tmp/haproxy_chroot/.well-known/acme-challenge/"
+	. $domainslist->Draw($a_domains)
+));
```
This option uses the fa-level-down icon to create a new entry, which is unintuitive and easy to miss. Copy the style used by rowhelper in various areas of the GUI and use a button with text

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_certificates_edit.php:
```diff
@@ -0,0 +1,432 @@
+$section->addInput(new \Form_StaticText(
+	'Actions list', 
+	"Used to restart webserver processes this certificate has been renewed<br/>Examples:<br/>/etc/rc.restart_webgui<br/>/usr/local/etc/rc.d/haproxy.sh restart".
+	$actionslist->Draw($a_actions)
```
This option uses the fa-level-down icon to create a new entry, which is unintuitive and easy to miss. Copy the style used by rowhelper in various areas of the GUI and use a button with text

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_accountkeys_edit.php:
```diff
@@ -0,0 +1,503 @@
+$section->addInput(new \Form_StaticText(
+	'', 
+	"<a id='btncreatekey' class='btn btn-sm btn-primary'>"
+		. "<i id='btncreatekeyicon' class='fa fa-check'></i> Create new account key</a>"
```
Create button should not use a checkmark, use fa-plus instead.

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_accountkeys_edit.php:
```diff
@@ -0,0 +1,503 @@
+$section->addInput(new \Form_StaticText(
+	'Acme account registration',
+	"<a id='btnregisterkey' class='btn btn-sm btn-primary'>"
+		. "<i id='btnregisterkeyicon' class='fa fa-check'></i> Register acme account key</a>"
```
Register button should not use a checkmark. Use an icon that implies account registration, like fa-user, fa-key, maybe fa-lock.

jim-p commented on security/pfSense-pkg-acme/files/usr/local/pkg/acme/acme.sh:
```diff
@@ -0,0 +1,4587 @@
+#!/usr/bin/env sh
```
This script needs the execute bit set.

PiBa-NL replied:
makefile sets -m 0755, that should be enough right?

jim-p replied:
I didn't notice that, but yeah it should be sufficient in both cases. I just noticed when looking at the diff that it was 644.

rbgarga replied:
@PiBa-NL INSTALL_DATA macro already set -m 0644, add it explicit is redundant. And for scripts you should use INSTALL_SCRIPT macro and remove -m 0755

PiBa-NL replied:
Now changed to use INSTALL_DATA/INSTALL_SCRIPT macros in the Makefile.

jim-p commented on security/pfSense-pkg-acme/files/usr/local/pkg/acme/acme_command.sh:
```diff
@@ -0,0 +1,66 @@
+#!/usr/local/bin/php -f
```
This script needs the execute bit set.

PiBa-NL replied:
makefile sets -m 0755, that should be enough right?

PiBa-NL replied:
Now changed to use INSTALL_DATA/INSTALL_SCRIPT macros in the Makefile.

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_accountkeys_edit.php:
```diff
@@ -0,0 +1,503 @@
+	/* Ensure that our pool names are unique */
+	for ($i=0; isset($config['installedpackages']['acme']['accountkeys']['item'][$i]); $i++) {
+		if (($_POST['name'] == $config['installedpackages']['acme']['accountkeys']['item'][$i]['name']) && ($i != $id)) {
+			$input_errors[] = "This pool name has already been used.  Pool names must be unique.";
```
"Pool name"? Shouldn't this be "Account Key Name"?

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_accountkeys_edit.php:
```diff
@@ -0,0 +1,503 @@
+	}
+
+	if($accountkey['name'] != "") {
+		$changedesc .= " modified pool: '{$accountkey['name']}'";
```
Pool?

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_certificates_edit.php:
```diff
@@ -0,0 +1,432 @@
+	/* Ensure that our pool names are unique */
+	for ($i=0; isset($config['installedpackages']['acme']['certificates']['item'][$i]); $i++) {
+		if (($_POST['name'] == $config['installedpackages']['acme']['certificates']['item'][$i]['name']) && ($i != $id)) {
+			$input_errors[] = "This pool name has already been used.  Pool names must be unique.";
```
Pool name?

jim-p commented on security/pfSense-pkg-acme/files/usr/local/www/acme/acme_certificates_edit.php:
```diff
@@ -0,0 +1,432 @@
+	}
+
+	if($certificate['name'] != "") {
+		$changedesc .= " modified pool: '{$certificate['name']}'";
```
Pool?

rbgarga commented:
I'll wait the changes for items pointed out by @jim-p before start reviewing

jim-p commented:
One more question: Why is `nsupdate` support commented out? That would be one of the best choices for a lot of users.

PiBa-NL commented:
Regarding nsupdate, i will add below run_depends item, would that make its binary available.?
`RUN_DEPENDS=	samba-nsupdate:dns/samba-nsupdate`

jim-p commented:
We already have nsupdate, it is what we use for RFC 2136 updates.
`: pkg which /usr/local/bin/nsupdate
/usr/local/bin/nsupdate was installed by package bind-tools-9.11.0P1_1`

PiBa-NL commented:
Ah ok, well then i think all comments have been addressed.

The buttons for generating/registering account key still change to a check-mark when successful, i think thats alright. I've added the nsupdate option (someone should test it..). Added a little more text. Removed the 'pool' texts (which came from copy/paste from haproxy package..)

 Lemme know if i missed something :).

rbgarga commented:
@PiBa-NL here is output of 'portlint -CN', please fix all warnings and fatal messages
```
WARN: /usr/home/garga/work/pfsense/freebsd-ports/security/pfSense-pkg-acme/pkg-plist: [32]: empty line found in plist.
WARN: /usr/home/garga/work/pfsense/freebsd-ports/security/pfSense-pkg-acme/pkg-plist: [33]: empty line found in plist.
WARN: /usr/home/garga/work/pfsense/freebsd-ports/security/pfSense-pkg-acme/pkg-plist: [34]: empty line found in plist.
WARN: /usr/home/garga/work/pfsense/freebsd-ports/security/pfSense-pkg-acme/pkg-plist: seems to have unnecessary blank lines at the last part.
WARN: /usr/home/garga/work/pfsense/freebsd-ports/security/pfSense-pkg-acme/pkg-descr: includes lines that exceed 80 characters.
FATAL: Makefile: [13]: use a tab (not space) after a variable name
WARN: Makefile: [15]: whitespace before end of line.
FATAL: Makefile: MAINTAINER address, PiBa-NL, does not appear to be a valid email address.
WARN: Makefile: COMMENT should begin with a capital, and end without a period
WARN: Makefile: Consider defining LICENSE.
```

rbgarga commented on security/pfSense-pkg-acme/Makefile:
```diff
@@ -0,0 +1,97 @@
+
+MAINTAINER=	PiBa-NL
+COMMENT=	pfSense package acme
+
```
Add LICENSE=	APACHE20 between COMMENT and USE_PHP leaving a blank line before and after

rbgarga commented on security/pfSense-pkg-acme/Makefile:
```diff
@@ -0,0 +1,97 @@
+MAINTAINER=	PiBa-NL
+COMMENT=	pfSense package acme
+
+USE_PHP= ftp
```
USE_PHP alone is deprecated. keep this line replacing space by TAB to separate = and value and add the following line before it:
```
USES=	php
```

rbgarga commented on security/pfSense-pkg-acme/Makefile:
```diff
@@ -0,0 +1,97 @@
+
+USE_PHP= ftp
+
+CONFLICTS=	
```
Remove blank CONFLICT block

rbgarga commented on security/pfSense-pkg-acme/Makefile:
```diff
@@ -0,0 +1,97 @@
+do-install:
+	${MKDIR} ${STAGEDIR}${PREFIX}/pkg
+	${MKDIR} ${STAGEDIR}${PREFIX}/pkg/acme
+	${MKDIR} ${STAGEDIR}${PREFIX}/pkg/acme/dnsapi
```
MKDIR macro is 'mkdir -p' so you just need to run it to last level and it will create parent directories, remove lines creating pkg and pkg/acme

rbgarga commented on security/pfSense-pkg-acme/Makefile:
```diff
@@ -0,0 +1,97 @@
+	${MKDIR} ${STAGEDIR}${PREFIX}/pkg
+	${MKDIR} ${STAGEDIR}${PREFIX}/pkg/acme
+	${MKDIR} ${STAGEDIR}${PREFIX}/pkg/acme/dnsapi
+	${MKDIR} ${STAGEDIR}${PREFIX}/www
```
Remove this line, next one will create this directory too

rbgarga commented on security/pfSense-pkg-acme/pkg-plist:
```diff
@@ -0,0 +1,34 @@
+www/acme/acme_generalsettings.php
+/etc/inc/priv/acme.priv.inc
+%%DATADIR%%/info.xml
+@dir /etc/inc/priv
```
Add @dir /etc/inc and remove all blank lines

jim-p commented:
I'm still checking this out but got sidetracked by other things I was working on. I did notice that in the Domain SAN list when you edit an entry the form fields only take up a little amount of space horizontally and it looks weird.

And it may be beyond the scope of this code, but I'd like to see the nsupdate method support an additional field for a domain suffix or fulldomain override. I can see people not wanting to add dynamic updates for their entire zone, but you could make a new dynamic zone or subdomain and CNAME the `_acme-challenge` record there. For example, dynamic zones happen in `.le.example.com`. -- so `_acme-challenge.server` would be a CNAME to `server.example.com.le.example.com` similar to what is mentioned at https://www.crc.id.au/using-centralised-management-with-lets-encrypt/ -- That may be something we can add on after this gets merged.

jim-p commented:
Screen cap of the weird formatting: http://i.imgur.com/p1JjE4I.png

jim-p commented:
Looks like maybe it's to a point where we could pull it in and hack on it some more as we go.

rbgarga commented:
Merged in a single commit to make it easier. Thanks!!!

979e112899dade6a3e00138e5d5ca563227e980d

OmgImAlexis commented:
@jim-p roughly how long until this commit will make it into master?

jim-p commented:
It's in master, but on master new packages only show up when a new snapshot is built. So the next snapshot will have it. It's possible as well that it wasn't built with the latest snapshot run and needs adjustments yet. So probably tomorrow sometime if it isn't up already
